OrbitCore/BpfTrace.cpp

```diff
@@ -262,3 +286,158 @@ void BpfTrace::CommandCallbackWithCallstacks(const std::string& a_Line)
+
+        // start capturing
+        LinuxPerfUtils::start_capturing(uprobe_fd);
+        LinuxPerfUtils::start_capturing(uretprobe_fd);
+    }
+
+    LinuxPerfEventProcessor event_buffer(std::make_unique<BpfTraceVisitor>());
+
+    bool new_events = false;
+
+    while( !(*a_ExitRequested) )
+    {
+        // Lets sleep a bit, such that we are not constantly reading from the buffers
+        // and thus wasting cpu time. 1000 microseconds are still small enought to 
+        // not have our buffers overflown and therefore loosing events.
+        if (!new_events)
+        {
+            usleep(10000);
+        }
+
+        new_events = false;
+
+        // read from all ring buffers, create events and store them in the event_queue
+        // TODO: implement a better scheduling strategy.
+        for (auto& pair : uprobe_ring_buffers)
+        {
+            auto& ring_buffer = pair.second;
+            int i = 0;
+            // read everything that is new
+            while (ring_buffer.HasNewData() && i < ROUND_ROBIN_BATCH_SIZE)
+            {
+                perf_event_header header{};
+                ring_buffer.ReadHeader(&header);
+
+                // perf_event_header::type contains the type of record,
+                // defined in enum perf_event_type in perf_event.h.
+                switch (header.type)
+                {
+                case PERF_RECORD_SAMPLE:
+                    {
+                        if (GParams.m_BpftraceCallstacks)
+                        {
+                            auto record = ring_buffer.ConsumeRecord<LinuxUprobeEventWithStack>(header);
+                            record.SetFunction(pair.first);
+                            event_buffer.Push(std::make_unique<LinuxUprobeEventWithStack>(std::move(record)));
+                        }
+                        else
+                        {
+                            auto record = ring_buffer.ConsumeRecord<LinuxUprobeEvent>(header);
+                            record.SetFunction(pair.first);
+                            event_buffer.Push(std::make_unique<LinuxUprobeEvent>(std::move(record)));
+                        }
+                        
+                    }
+                    break;
+                 case PERF_RECORD_LOST:
+                    {
+                        auto lost = ring_buffer.ConsumeRecord<LinuxPerfLostEvent>(header);
+                        PRINT("Lost %u Events\n", lost.Lost());
+                    }
+                    break;
+                default:
+                    PRINT("Unexpected Perf Sample Type: %u", header.type);
+                    ring_buffer.SkipRecord(header);
+                    break;
+                }
+            }
+        }
+
+        for (auto& pair : uretprobe_ring_buffers)
+        {
+            auto& ring_buffer = pair.second;
+            int i = 0;
+            // read everything that is new
+            while (ring_buffer.HasNewData() && i < ROUND_ROBIN_BATCH_SIZE)
```

Comment:
Looks like "i" is not incremented.

Reply:
See [comment](https://github.com/pierricgimmig/orbitprofiler/pull/89#discussion_r354154537) above.